crates/wasm-smith/src/core.rs

```diff
@@ -968,7 +979,7 @@ impl Module {
 
         if self.config.gc_enabled && !self.types.is_empty() && u.arbitrary()? {
             let type_ref_limit = u32::try_from(self.types.len()).unwrap();
-            let idx = u.int_in_range(0..=type_ref_limit)?;
+            let idx = u.int_in_range(0..=type_ref_limit - 1)?;
```

Comment:
I think this was not `-1` to allow type references to `self` when we are constructing a type, which isn't yet in `self.types`. But that only works when we are constructing a new type to add to `self.types`, not for arbitrary callers. Maybe it makes sense to take a parameter to control this behavior?

Reply:
I opted to refactor the various infrastructure here a bit more to ensure that this is handled throughout all type generation, mind taking a look again?